### PR TITLE
avm2: Implement some of FileReference including browse/save

### PR DESCRIPTION
**Pull request by evilpie**
Makes file saving ("downloading") and file opening ("uploading") work in Scratch 2 (#12613).

Edit: Uploading in scratch doesn't work yet, because I removed support for `FileReferenceList.browse` from this PR. New PR coming soon.

### CODE REVIEW

File: core/src/avm2/object/file_reference_object.rs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    fn as_file_reference(&self) -> Option<FileReferenceObject<'gc>> {
+        Some(*self)
+    }
+}
+
+impl<'gc> FileReferenceObject<'gc> {
+    pub fn init_from_dialog_result(&self, result: Box<dyn FileDialogResult>) -> FileReference {
+        println!("init_from_dialog_result");
```
Comment (Lord-McSweeney) [inline core/src/avm2/object/file_reference_object.rs:60]: Leftover `println!`

**Comment by danielhjacobs** [comment]
Not sure if you saw it since it was opened so long it's buried but you may want to compare this to https://github.com/ruffle-rs/ruffle/pull/7637.

**Comment by evilpie** [comment]
> Not sure if you saw it since it was opened so long it's buried but you may want to compare this to #7637.

Ah nice. I hadn't seen this. This would probably be a good basis for implementing FileReferenceList in avm2. For now the partial implementation is good enough.

File: core/src/loader.rs [hunk 2]
```diff
@@ -704,6 +744,28 @@ pub enum Loader<'gc> {
...
+
+    SaveFileDialog {
+        /// The handle to refer to this loader instance.
+        #[collect(require_static)]
+        self_handle: Option<Handle>,
+
+        /// The target AVM2 object to select a file path from.
+        // XXXXXX
+
+        // FIXME
```
Comment (Lord-McSweeney) [inline core/src/loader.rs:765]: remaining FIXME?

File: core/src/avm2/globals/flash/net/file_reference.rs [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+    activation: &mut Activation<'_, 'gc>,
+    this: Object<'gc>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    let this = this.as_file_reference().unwrap();
+
+    let name = match *this.file_reference() {
+        FileReference::None => return Err(make_error_2037(activation)),
+        FileReference::FileDialogResult(ref dialog_result) => {
+            // XXX Value::Null?
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/net/file_reference.rs:39]: remaining FIXME?

File: core/src/avm2/globals/flash/net/file_reference.rs [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+        FileReference::FileDialogResult(ref dialog_result) => {
+            // XXX not sure how test this
+            dialog_result.size().unwrap_or(0)
+        }
+    };
+
+    Ok(Value::Number(size as f64))
+}
+
+fn make_error_2097<'gc>(activation: &mut Activation<'_, 'gc>) -> Error<'gc> {
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/net/file_reference.rs:67]: This should also be moved into `error.rs`

**Review by Aaron1011 — APPROVED** [review]

**Review by Dinnerbone — APPROVED** [review]
LGTM! Thank you!

**Pull request merged**